napi-inl.h

```diff
@@ -1054,99 +1056,75 @@ inline Napi::ArrayBuffer TypedArray::ArrayBuffer() const {
-  return New(env, elementLength, arrayBuffer, 0);
+  return New(env, elementLength, arrayBuffer, 0, type);
 }
 
-template <typename T, napi_typedarray_type A>
-inline TypedArray_<T,A> TypedArray_<T,A>::New(napi_env env,
-                                              size_t elementLength,
-                                              Napi::ArrayBuffer arrayBuffer,
-                                              size_t bufferOffset) {
+template <typename T>
+inline TypedArrayOf<T> TypedArrayOf<T>::New(napi_env env,
+                                            size_t elementLength,
+                                            Napi::ArrayBuffer arrayBuffer,
+                                            size_t bufferOffset,
+                                            napi_typedarray_type type) {
   napi_value value;
   napi_status status = napi_create_typedarray(
-    env, A, elementLength, arrayBuffer, bufferOffset, &value);
+    env, type, elementLength, arrayBuffer, bufferOffset, &value);
   if (status != napi_ok) throw Error::New(env);
 
-  return TypedArray_<T,A>(env, value, elementLength, reinterpret_cast<T*>(arrayBuffer.Data()));
+  return TypedArrayOf<T>(env, value, type, elementLength, reinterpret_cast<T*>(arrayBuffer.Data()));
 }
 
-template <typename T, napi_typedarray_type A>
-inline TypedArray_<T,A>::TypedArray_() : TypedArray(), _data(nullptr) {
+template <typename T>
+inline TypedArrayOf<T>::TypedArrayOf() : TypedArray(), _data(nullptr) {
 }
 
-template <typename T, napi_typedarray_type A>
-inline TypedArray_<T,A>::TypedArray_(napi_env env, napi_value value)
-  : TypedArray(env, value, A, 0), _data(nullptr) {
+template <typename T>
+inline TypedArrayOf<T>::TypedArrayOf(napi_env env, napi_value value)
+  : TypedArray(env, value), _data(nullptr) {
   napi_status status = napi_get_typedarray_info(
-    _env, _value, nullptr, &_length, reinterpret_cast<void**>(&_data), nullptr, nullptr);
+    _env, _value, &_type, &_length, reinterpret_cast<void**>(&_data), nullptr, nullptr);
   if (status != napi_ok) throw Error::New(_env);
 }
 
-template <typename T, napi_typedarray_type A>
-inline TypedArray_<T,A>::TypedArray_(napi_env env, napi_value value, size_t length, T* data)
-  : TypedArray(env, value, A, length), _data(data) {
+template <typename T>
+inline TypedArrayOf<T>::TypedArrayOf(napi_env env,
+                                     napi_value value,
+                                     napi_typedarray_type type,
+                                     size_t length,
+                                     T* data)
+  : TypedArray(env, value, type, length), _data(data) {
+  if (!(type == TypedArrayTypeForPrimitiveType<T>() ||
+      (type == napi_uint8_clamped_array && std::is_same<T, uint8_t>::value))) {
+    throw TypeError::New(env, "Array type must match the template parameter. "
+    " (Uint8 arrays may optionally have the \"clamped\" array type.)");
```

Comment:
I'll remove the extra space.

I didn't refer to `Uint8Array` specifically because that's just a typedef; Some people may prefer to just use `TypedArrayOf<uint8_t>` instead.